platforms/documentation/docs/src/docs/userguide/best-practices/bp_general.adoc

```diff
@@ -15,4 +15,36 @@
 [[best_practices_general]]
 = General Gradle Best Practices
 
-Add content here.
+== Do Not Use Internal APIs
+
+Do not use APIs that are inside an `internal` Java package or contain `Internal` or `Impl`
+in the name of the type.
+
+=== Explanation
+
+Using internal APIs is inherently risky and can cause significant problems during upgrades.
+Gradle and many plugins (such as Android Gradle Plugin and Kotlin Gradle Plugin) treat internal
+APIs, such as those within internal packages, as subject to breaking changes, even during minor
```

Comment:
```suggestion
APIs, such as those within internal packages, as subject to unannounced breaking changes during any new Gradle release, even during minor
```

I think we need to really hammer the point that there are no backwards or forwards compatibility guarantees here.